### PR TITLE
Adding preconditioner to idrs

### PR DESCRIPTION
**Pull request by deltaeecs**
Here I added Pl preconditioner to idrs add tested it with IncompleteLU preconditioner.

### CODE REVIEW

**Comment by mschauer** [comment]
Nice! Did you follow literature or is it just clear how to add the preconditioner here? Pinging @astudillor, can you also have a cursory look?

File: src/idrs.jl [hunk 1]
```diff
@@ -78,8 +80,8 @@ end
 end
 
 function idrs_method!(log::ConvergenceHistory, X, A, C::T,
-    s::Number, abstol::Real, reltol::Real, maxiter::Number; smoothing::Bool=false, verbose::Bool=false
-    ) where {T}
+    s::Number, Pl::precT, abstol::Real, reltol::Real, maxiter::Number; smoothing::Bool=false, verbose::Bool=false
+    ) where {T, precT}
```
Comment (mschauer) [inline src/idrs.jl:84]: I think you never use `precT`, so you don't need that type parameter in the method signature

File: Project.toml [hunk 2]
```diff
@@ -3,6 +3,7 @@ uuid = "42fd0dbc-a981-5370-80f2-aaf504508153"
 version = "0.9.0"
 
 [deps]
+IncompleteLU = "40713840-3770-5561-ab4c-a76e7d0d7895"
```
Comment (mschauer) [inline Project.toml:6]: We have a file for test dependencies, https://github.com/JuliaLinearAlgebra/IterativeSolvers.jl/blob/master/test/Project.toml, we only need that dependency for the test, right?

File: src/idrs.jl (hunk 1 above)
Comment (deltaeecs) [reply src/idrs.jl:84]: Yes, I just followed the "gmres.jl" and added "precT". Maybe it will be useful in the future?

File: Project.toml (hunk 2 above)
Comment (deltaeecs) [reply Project.toml:6]: I just checked the "gmres.jl" and found it use LU instead of the IncompleteLU I added in the dependency，shall I removed it?

Comment (mschauer) [reply Project.toml:6]: It makes sense to have a non-exact preconditioner in the test, or? Perhaps just move the 
```
IncompleteLU = "40713840-3770-5561-ab4c-a76e7d0d7895"
```
line to https://github.com/JuliaLinearAlgebra/IterativeSolvers.jl/blob/master/test/Project.toml

**Comment by deltaeecs** [comment]
> Nice! Did you follow literature or is it just clear how to add the preconditioner here? Pinging @astudillor, can you also have a cursory look?

hahaha
I followed the Matlab Code referred in the docs because I found idrs performs better  than gmres in my work~

File: Project.toml (hunk 2 above)
Comment (deltaeecs) [reply Project.toml:6]: ok

**Comment by mschauer** [comment]
Haha, you can also compare with https://github.com/astudillor/idrs/blob/master/idrs/idrs.py

File: test/idrs.jl [hunk 3]
```diff
@@ -42,6 +43,25 @@ end
     @test norm(A * x - b) / norm(b) ≤ reltol
 end
 
+@testset "SparseMatrixCSC{$T, $Ti}" for T in (Float64, ComplexF64), Ti in (Int64, Int32)
```
Comment (mschauer) [inline test/idrs.jl:45]:
```suggestion
@testset "SparseMatrixCSC{$T, $Ti} with preconditioner" for T in (Float64, ComplexF64), Ti in (Int64, Int32)
```

**Comment by deltaeecs** [comment]
> Haha, you can also compare with https://github.com/astudillor/idrs/blob/master/idrs/idrs.py

Thank you for your help~
I think they are the same after checked the idrs.py~
I'm so happy since it's my first time to do a little work on git~

**Comment by mschauer** [comment]
Great. Can I ask on what type of problem you are working such that IDR(s) beats GMRES?

**Comment by deltaeecs** [comment]
> Great. Can I ask on what type of problem you are working such that IDR(s) beats GMRES?

It's a badly conditioned matrix equation on Computational Electromagneticcalled EFIE~

**Review by mschauer — APPROVED** [review]

**Comment by deltaeecs** [comment]
> > Great. Can I ask on what type of problem you are working such that IDR(s) beats GMRES?
> 
> It's a badly conditioned matrix equation on Computational Electromagneticcalled EFIE~

Well it seems to be a problem of problem of setting on preconditioner. GMRES is still the best solver after I corrected it.

**Pull request merged**